services/wallet/thirdparty/fourbyte/client.go

```diff
@@ -0,0 +1,91 @@
+package fourbyte
+
+import (
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"net/http"
+	"regexp"
+	"strings"
+	"time"
+
+	"github.com/ethereum/go-ethereum/accounts/abi"
+	"github.com/status-im/status-go/services/wallet/thirdparty"
+)
+
+type Signature struct {
+	ID   int    `json:"id"`
+	Text string `json:"text_signature"`
+}
+
+type SignatureList struct {
+	Count   int         `json:"count"`
+	Results []Signature `json:"results"`
+}
+
+type Client struct {
+	client *http.Client
+}
+
+func NewClient() *Client {
+	return &Client{client: &http.Client{Timeout: time.Minute}}
+}
+
+func (c *Client) DoQuery(url string) (*http.Response, error) {
+	resp, err := c.client.Get(url)
+
+	if err != nil {
+		return nil, err
+	}
+	return resp, nil
+}
+
+func (c *Client) Run(data string) (*thirdparty.DataParsed, error) {
+	methodSigData := data[2:10]
+	url := fmt.Sprintf("https://www.4byte.directory/api/v1/signatures/?hex_signature=%s", methodSigData)
+	resp, err := c.DoQuery(url)
+	if err != nil {
+		return nil, err
+	}
+	defer resp.Body.Close()
+
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return nil, err
+	}
+	var signatures SignatureList
+	err = json.Unmarshal(body, &signatures)
+	if err != nil {
+		return nil, err
+	}
+
+	if signatures.Count == 0 {
+		return nil, err
+	}
+
+	rgx := regexp.MustCompile(`\((.*?)\)`)
+	signature := signatures.Results[0].Text
```

Comment:
(My understanding is) the API allows retrieving function signatures filtered by different criteria. For some of those, it makes sense to receive several results (filtering by text signature), but since we're using the exact hash, if there's a match it should be unique.

Reply:
@alaibe there are some edge cases that might happen with retrieving just the first result. for example:  the `balanceOf` function signature: https://www.4byte.directory/api/v1/signatures/?hex_signature=0x70a08231. 

Notice how using only the first result, we would display the function: `watch_tg_invmru_119a5a98(address,uint256,uint256)`, although most of the time an user would expect to see `balanceOf` instead. I think the heuristics for determining which function should be used must be changed, for example,  by using the result with the oldest id,  or attempting to decode the `data` with all the possible signatures until there's a match (however, i don't know if `method.Inputs.UnpackIntoMap` would fail if the data does not match the expected signature)

Interestingly enough, metamask only takes into account 4bytes directory when there's just 1 result: https://github.com/MetaMask/metamask-extension/blob/e3ea4f2cd044d48c61b6a35ef206fa942f3b43d1/ui/helpers/utils/transactions.util.js#L40

Reply:
Indeed i can try to check if the unpack into map fail or not first, if it does, I will then use it. If not, as u said I will use oldest ID
Good call!